lib/async_hooks.js

```diff
@@ -197,6 +200,98 @@ class AsyncResource {
   }
 }
 
+function getResource(id) {
+  const weak = idResourcePairing.get(id);
+  if (!weak) {
+    return undefined;
+  }
+  return weak.get();
+}
+
+function getCurrentResource() {
+  const id = executionAsyncId();
+  return getResource(id);
+}
+
+const idResourcePairing = new Map();
+const contexts = new Set();
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // Save new resource.
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResource = getResource(triggerAsyncId);
+    if (triggerResource) {
+      contexts.forEach((context) => {
+        context._propagate(resource, triggerResource);
+      });
```

Comment:
pleas use a standard for(;;) loop, this is very perf sensitive.

Reply:
sure! Will do

Reply:
Done!